### PR TITLE
EDGE-694-immediate-master-listen-return when mg_stop is called

### PR DESCRIPTION
**Pull request by vvmimik**

### CODE REVIEW

File: src/civetweb.c [hunk 1]
```diff
@@ -16633,6 +16643,15 @@ master_thread_run(void *thread_func_param)
 					accept_new_connection(&ctx->listening_sockets[i], ctx);
 				}
 			}
+
+			if(ctrlidx && (pfd[i].revents & POLLIN) && (pfd[i].fd == ctx->listen_ctrl_sd)){
+				int rcvd = 0 ;
+				/* At this time not interested in what is recvd,
+				   but rather a way to wake up from poll */
+				recv(ctx->listen_ctrl_sd,&rcvd,sizeof(rcvd),0);
+				close(ctx->listen_ctrl_sd);
```
Comment (solvexyz) [inline src/civetweb.c]: I think we will have a resource leak/memory leak if terminate the civet instance using signal.  Then close will not happen.   I prefer move this to mg_stop (close and assign invalid).

File: src/civetweb.c [hunk 2]
```diff
@@ -16811,6 +16830,17 @@ mg_stop(struct mg_context *ctx)
 	/* Set stop flag, so all threads know they have to exit. */
 	ctx->stop_flag = 1;
 
+	if(ctx->listen_ctrl_sd > 0){
```
Comment (solvexyz) [inline src/civetweb.c]: Code formatting

File: src/civetweb.c [hunk 3]
```diff
@@ -16811,6 +16830,17 @@ mg_stop(struct mg_context *ctx)
...
+		struct sockaddr_in to;
+		int l = sizeof(to);
+		memset(&to, 0, sizeof(to));
+		if(!getsockname(ctx->listen_ctrl_sd, (struct sockaddr*)&to, &l)){
```
Comment (solvexyz) [inline src/civetweb.c]: code formatting

File: src/civetweb.c [hunk 4]
```diff
@@ -17060,6 +17090,26 @@ mg_start(const struct mg_callbacks *callbacks,
 		return NULL;
 	}
 
+	/* listen_ctrl_sd is used by mg_stop to make master thread
+	   break out safely from i/o wait in poll instead of relying on 
+	   frequent poll timeout to wake up and check the stop flag.
+	*/
+	ctx->listen_ctrl_sd = socket(AF_INET, SOCK_DGRAM, 0);
```
Comment (solvexyz) [inline src/civetweb.c:17104]: we need to handle listen_ctrl_sd error case by returning NULL like rest of the error in mg_start

File: src/civetweb.c [hunk 5]
```diff
@@ -17060,6 +17090,26 @@ mg_start(const struct mg_callbacks *callbacks,
...
+	if(ctx->listen_ctrl_sd > 0) {
```
Comment (solvexyz) [inline src/civetweb.c]: code formatting

File: src/civetweb.c [hunk 6]
```diff
@@ -17060,6 +17090,26 @@ mg_start(const struct mg_callbacks *callbacks,
...
+		struct sockaddr_in saddr;
+		memset(&saddr, 0, sizeof(saddr));
+		saddr.sin_family = AF_INET;
+		saddr.sin_port = htons(0);
+		saddr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+		fcntl(ctx->listen_ctrl_sd, F_SETFD, FD_CLOEXEC);
+		set_non_blocking_mode(ctx->listen_ctrl_sd);
+		if(bind(ctx->listen_ctrl_sd, (struct sockaddr *)&saddr,
+                                                   sizeof(saddr)) < 0) {
+			close(ctx->listen_ctrl_sd);
```
Comment (solvexyz) [inline src/civetweb.c]: We need to handle error by returning null, and make sure other resources are properly cleared.

**Comment by vvmimik** [comment]
Hi Jermey
Thank you for reviewing the changes.

I will explain in detail as why it was made
that way when I get to office today late afternoon after my heart appointment.

The only resource we use is creation of a socket descriptor which is closed on being signalled or by default when the process is terminated by signal. Should you need, I can move the close to termination function.

At this time,  failure of this ctrl udp socket and other errors does not prevent edge from running successfully.

So, only if everything is successful, we use this socket  to signal listen thread waiting on poll to break immediately else we rely on 5 second timeout to comeout of poll.

Should you need , I can check errors for the calls made and return from start , which is only needed if we are going to remove timeout completely. Currently if we get errors , we do not use this socket.

Thanks
Regards
Varadhan

________________________________

**Pull request merged**